jax/interpreters/masking.py

```diff
@@ -107,77 +107,93 @@ def eval_poly_shape(shape, values_dict):
     return Poly({mon: -coeff for mon, coeff in self.items()})
 
-  def __mul__(self, other):
+  def __mul__(self, other: 'Size') -> 'Poly':
     other = _ensure_poly(other)
-    coeffs = {}
+    coeffs: Dict[Mon, int] = {}
     for (mon1, coeff1), (mon2, coeff2) in product(self.items(), other.items()):
       mon = mon1 * mon2
       coeffs[mon] = coeffs.get(mon, 0) + coeff1 * coeff2
     return Poly(coeffs)
 
-  def __rmul__(self, other):
+  def __rmul__(self, other: 'Size') -> 'Poly':
     return self * other  # multiplication commutes
 
-  def __radd__(self, other):
+  def __radd__(self, other: 'Size') -> 'Poly':
     return self + other  # addition commutes
 
-  def __rsub__(self, other):
+  def __rsub__(self, other: 'Size') -> 'Poly':
     return _ensure_poly(other) - self
 
-  def __floordiv__(self, divisor):
-    q, _ = divmod(self, divisor)  # pytype: disable=wrong-arg-types
+  def __floordiv__(self, divisor: 'Size') -> 'Poly':
+    q, _ = divmod(self, divisor)
     return q
 
-  def __mod__(self, divisor):
-    _, r = divmod(self, divisor)  # pytype: disable=wrong-arg-types
+  def __mod__(self, divisor: 'Size') -> int:
+    _, r = divmod(self, divisor)
     return r
 
-  def __divmod__(self, divisor):
-    if self.is_constant:
-      return divmod(int(self), divisor)
-    else:
-      def divided(count):
-        q, r = divmod(count, divisor)
-        if r != 0:
-          raise ValueError('shapecheck  and masking currently only support '
-                           'strides that exactly divide the strided axis '
-                           'length.')
-        return q
-
-      return Poly(
-        {k: coeff // divisor if k.degree == 0 else divided(coeff)
-         for k, coeff in self.items()}), self.get(Mon(), 0) % divisor
+  def __divmod__(self, divisor: 'Size') -> Tuple['Poly', int]:
+    """
+    Floor division with remainder (divmod) generalized to polynomials.
+    To allow ensuring consistency with integer divmod,
+    the polynomial division remainder divided by the divisor must be constant.
+    :return: Quotient resulting from polynomial division and integer remainder.
+    """
+    divisor = _ensure_poly(divisor)
+    quotient, remainder = self._polynomial_divmod(divisor)
+    if not remainder.is_constant:
+      raise ValueError(f"Stride {divisor} for size {self} must have constant "
+                       f"remainder, but was {remainder}.")
+    if remainder != 0 and not divisor.is_constant:
+      raise ValueError(f"Polymorphic stride {divisor} must divide size {self}.")
+    return quotient, int(remainder)
+
+  def __rdivmod__(self, dividend: 'Size') -> Tuple['Poly', int]:
+    return divmod(_ensure_poly(dividend), self)
+
+  def _polynomial_divmod(self, divisor: 'Poly') -> Tuple['Poly', 'Poly']:
+    """Multivariate polynomial division, returning quotient and remainder."""
+    dmon, dcount = divisor._leading_term
+    dividend, quotient, remainder = self, _ensure_poly(0), _ensure_poly(0)
+    while dividend != 0:
```

Comment:
I think it would help if you wrote down the loop invariant that dividend = divisor * quotient + remainder. 

I am trying to convince myself that this is "complete". E.g., if the leading term of the dividend and the divisor are not the same, then the monomial division will return None, and the entire leading term of the dividend is added to the remainder. I guess I am not sure what means the leading term. If it were the one with the highest degree, I could see how this terminates. But otherwise, why does it terminate?

Reply:
@gnecula I described the invariant in the code. 

The leading terms is the highest degree term that comes first lexicographically. (The monomials are ordered in [graded reverse lexicographical order](https://en.wikipedia.org/wiki/Monomial_order#Graded_reverse_lexicographic_order): First by total degree and then lexicographically to break ties.)